File: sdks/python-sdk/agentlabs/_internals/http.py
from typing import Any, Optional, TypedDict
import requests

class HttpApiConfig(TypedDict):
    agentlabs_url: str;
    project_id: str;
    secret: str;

class HttpApi:
    def __init__(self, agentlabs_url: str, project_id: str, secret: str) -> None:
        self.agentlabs_url = agentlabs_url
        self.project_id = project_id
        self.secret = secret
        self.client = requests.Session()
        self.client.params = {
            "projectId": self.project_id
        }
        self.client.headers.update({
            "x-agentlabs-sdk-secret": self.secret,
            "x-agentlabs-project-id": self.project_id,
        })


    def upload(self, path: str, filename: str, data: bytes, mime_type: Optional[str]) -> Any:
        mime_type = mime_type or "application/octet-stream"
        response = self.client.post(self._make_url(path), files={
            "file": (filename, data, mime_type)
        })
        response.raise_for_status()

        return response.json()

    def _make_url(self, path: str):
        return f"{self.agentlabs_url}/api{path}"

File: sdks/node-sdk/src/http.ts
import { ProjectConfig } from './project';

export interface UploadSettings {
	filename?: string;
	mimeType?: string;
}

export class HttpApi {
	constructor(private readonly projectConfig: ProjectConfig) {}

	private async fetchApi(path: string, options: RequestInit = {}): Promise<Response> {
		options.headers = {
			...options.headers,
			'x-agentlabs-sdk-secret': this.projectConfig.secret,
			'x-agentlabs-project-id': this.projectConfig.projectId,
		}

		const response = await fetch(`${this.projectConfig.url}/api${path}`, options);
		

		if (!response.ok) {
			throw new Error(`HTTP request failed: ${response.status} ${response.statusText}`);
		}

		return response;
	}

	private async apiFetchJson<T>(path: string, options: RequestInit = {}): Promise<T> {
		const response = await this.fetchApi(path, options);

		return response.json();
	}

	private async apiFetchBinary(path: string, options: RequestInit = {}): Promise<Buffer> {
		const response = await this.fetchApi(path, options);
		const abuf = await response.arrayBuffer();

		return Buffer.from(abuf);
	}

	async get<T>(path: string): Promise<T> {
		return this.apiFetchJson(path);
	}

	async download(path: string): Promise<Buffer> {
		return this.apiFetchBinary(path);
	}

	async post<T>(path: string, body: any): Promise<T> {
		return this.apiFetchJson(path, {
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
			},
			body: JSON.stringify(body),
		});
	}

	async upload<T>(path: string, buffer: Buffer, settings: UploadSettings = {}): Promise<T> {
		const formData = new FormData();
		const blob = new Blob([buffer], { type: settings?.mimeType });

		formData.append('file', blob, settings?.filename);

		return this.apiFetchJson(path, {
			method: 'POST',
			body: formData,
		});
	}
}
